src/main.ts

```diff
@@ -576,7 +588,9 @@ function generateDecode(ctx: Context, fullName: string, messageDesc: DescriptorP
       readSnippet = code`${type}.decode(reader, reader.uint32()).value`;
     } else if (isTimestamp(field)) {
       const type = basicTypeName(ctx, field, { keepValueType: true });
-      readSnippet = code`${utils.fromTimestamp}(${type}.decode(reader, reader.uint32()))`;
+      readSnippet = options.useDate
```

Comment:
Hm, I think it might be a simpler change to leave this line as it was, and then change line 589's "if isTimestamp(field)" to include "&& useDate". 

That way we'll only get into this `else if` if its timestamp & use date is true, otherwise we'll go into the `isMessage(field)` block next, which I think should handle `Timestamp`'s just fine, i.e. it will basically do that `${type}.decode` that currently you have re-typed-out as a the `:` case.

...could probably do that for most of these `if isTimestamp(field)` changes?

Reply:
I wonder if `isTimestamp` should be renamed to `isTimestampAsDate`, and have it take both `options` & `field`, because (...in theory...?) everywhere that `isTimestamp` is called today, we really mean "...so that we can do special `Date` handling".

I guess the JSON serialization is the exception to that rule, where we want to detect `Timestamp` and do something special even if its not a `Date`.

Reply:
OK, I've simplified the code where I can and updated the codegen. Since we need `isTimestamp` to be separable from `options.useDate` for the JSON serialisation, I guess we could add a wrapper function that checks both, but I don't think that adds much value.